gecko/glue/MediaSource.cpp

```diff
@@ -0,0 +1,134 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "MediaSource.h"
+
+#if MOZ_AV1
+#include "AOMDecoder.h"
+#endif
+#include "DecoderTraits.h"
+#include "MediaContainerType.h"
+#include "MediaMIMETypes.h"
+#include "mozilla/Logging.h"
+#include "mozilla/Preferences.h"
+#include "nsThreadManager.h"
+
+mozilla::LogModule* GetMediaSourceLog()
+{
+  static mozilla::LazyLogModule sLogModule("MediaSource");
+  return sLogModule;
+}
+
+#define MSE_DEBUG(arg, ...) MOZ_LOG(GetMediaSourceLog(), mozilla::LogLevel::Debug, ("MediaSource(%p)::%s: " arg, this, __func__, ##__VA_ARGS__))
+
+namespace mozilla {
+
+// Returns true if we should enable MSE webm regardless of preferences.
+// 1. If MP4/H264 isn't supported:
+//   * Windows XP
+//   * Windows Vista and Server 2008 without the optional "Platform Update Supplement"
+//   * N/KN editions (Europe and Korea) of Windows 7/8/8.1/10 without the
+//     optional "Windows Media Feature Pack"
+// 2. If H264 hardware acceleration is not available.
+// 3. The CPU is considered to be fast enough
+static bool
+IsWebMForced(DecoderDoctorDiagnostics* aDiagnostics)
+{
+  bool mp4supported =
+    DecoderTraits::IsMP4SupportedType(MediaContainerType(MEDIAMIMETYPE("video/mp4")),
+                                      aDiagnostics);
+  /* TODO (gecko-media) This requires importing mozilla/gfx/gfxVars.h and
+   *                    dependencies
+  bool hwsupported = gfx::gfxVars::CanUseHardwareVideoDecoding();
+#ifdef MOZ_WIDGET_ANDROID
+  return !mp4supported || !hwsupported || VP9Benchmark::IsVP9DecodeFast() ||
+         java::HardwareCodecCapabilityUtils::HasHWVP9();
+#else
+  return !mp4supported || !hwsupported || VP9Benchmark::IsVP9DecodeFast();
+#endif
+  */
+  return !mp4supported;
+}
+
+namespace dom {
+
+MediaSource::MediaSource(GeckoMediaSourceImpl aImpl)
+  : mImpl(aImpl)
+{
+}
+
+MediaSource::~MediaSource()
+{
+  if (mImpl.mContext && mImpl.mFree) {
```

Comment:
If you ever find yourself in future dispatching runnables to call callback functions, you'll need to ensure that you annull these callback functions in your shutdown. Otherwise at the point in time when we to shutdown the MediaSource, there may be runnables waiting to run that when run will call back over to Rust code that's been deallocated because Rust thinks everything is shutdown, and we'll crash. I hit this with the GeckoMediaDecoderOwner. You're not vulnerable to that yet, but it's a gotcha to look out for.